File: archive/sdk/python/paw/modules/dex.py
"""DEX module for trading operations."""

from typing import List, Optional
from ..types import (
    Pool,
    PoolParams,
    SwapParams,
    AddLiquidityParams,
    RemoveLiquidityParams,
    TxResult,
    GasOptions
)


class DexModule:
    """DEX module for decentralized exchange operations."""

    def __init__(self, client):
        """Initialize DEX module."""
        self.client = client

    async def create_pool(
        self,
        creator: str,
        params: PoolParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Create a new liquidity pool."""
        message = {
            "@type": "/paw.dex.v1.MsgCreatePool",
            "creator": creator,
            "token_a": params.token_a,
            "token_b": params.token_b,
            "amount_a": params.amount_a,
            "amount_b": params.amount_b
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def swap(
        self,
        sender: str,
        params: SwapParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Swap tokens in a pool."""
        message = {
            "@type": "/paw.dex.v1.MsgSwap",
            "sender": sender,
            "pool_id": params.pool_id,
            "token_in": params.token_in,
            "amount_in": params.amount_in,
            "min_amount_out": params.min_amount_out,
            "recipient": params.recipient or sender
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def add_liquidity(
        self,
        sender: str,
        params: AddLiquidityParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Add liquidity to a pool."""
        message = {
            "@type": "/paw.dex.v1.MsgAddLiquidity",
            "sender": sender,
            "pool_id": params.pool_id,
            "amount_a": params.amount_a,
            "amount_b": params.amount_b,
            "min_shares": params.min_shares
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def remove_liquidity(
        self,
        sender: str,
        params: RemoveLiquidityParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Remove liquidity from a pool."""
        message = {
            "@type": "/paw.dex.v1.MsgRemoveLiquidity",
            "sender": sender,
            "pool_id": params.pool_id,
            "shares": params.shares,
            "min_amount_a": params.min_amount_a,
            "min_amount_b": params.min_amount_b
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def get_pool(self, pool_id: str) -> Optional[Pool]:
        """Get pool by ID."""
        try:
            data = await self.client.get(f"/paw/dex/v1/pools/{pool_id}")
            pool_data = data.get("pool")
            if pool_data:
                return Pool(**pool_data)
            return None
        except Exception:
            return None

    async def get_all_pools(self) -> List[Pool]:
        """Get all pools."""
        try:
            data = await self.client.get("/paw/dex/v1/pools")
            pools = data.get("pools", [])
            return [Pool(**p) for p in pools]
        except Exception:
            return []

    def calculate_swap_output(
        self,
        amount_in: str,
        reserve_in: str,
        reserve_out: str,
        swap_fee: str = "0.003"
    ) -> str:
        """Calculate swap output amount."""
        amount_in_int = int(amount_in)
        reserve_in_int = int(reserve_in)
        reserve_out_int = int(reserve_out)
        fee_int = int(float(swap_fee) * 10000)

        amount_in_with_fee = (amount_in_int * (10000 - fee_int)) // 10000
        numerator = amount_in_with_fee * reserve_out_int
        denominator = reserve_in_int + amount_in_with_fee

        return str(numerator // denominator)

    def calculate_price_impact(
        self,
        amount_in: str,
        reserve_in: str,
        reserve_out: str
    ) -> float:
        """Calculate price impact percentage."""
        amount_out = self.calculate_swap_output(amount_in, reserve_in, reserve_out, "0")
        price_before = float(reserve_out) / float(reserve_in)
        price_after = float(amount_out) / float(amount_in)

        return abs((price_after - price_before) / price_before) * 100

File: archive/sdk/javascript/src/modules/dex.ts
import { PawClient } from '../client';
import {
  Pool,
  PoolParams,
  SwapParams,
  AddLiquidityParams,
  RemoveLiquidityParams,
  TxResult,
  GasOptions,
  ModuleDisabledError
} from '../types';

/**
 * DEX module for PAW blockchain
 *
 * ⚠️ WARNING: The DEX module is DISABLED in paw-mvp-1 testnet.
 * All query methods will return empty results.
 * All transaction methods will throw ModuleDisabledError.
 *
 * To enable: Submit a governance proposal to enable the DEX module.
 */
export class DexModule {
  private static MODULE_NAME = 'DEX';
  private moduleEnabled: boolean = false;

  constructor(private client: PawClient) {}

  /**
   * Check if DEX module is enabled on the chain
   */
  async isEnabled(): Promise<boolean> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');
      const response = await fetch(`${restEndpoint}/paw/dex/v1/params`);

      if (response.ok) {
        const data = await response.json() as { params?: { enabled?: boolean } };
        this.moduleEnabled = data?.params?.enabled ?? false;
        return this.moduleEnabled;
      }
      return false;
    } catch {
      return false;
    }
  }

  private assertEnabled(): void {
    // Always throw for now since DEX is disabled in testnet
    throw new ModuleDisabledError(DexModule.MODULE_NAME);
  }

  /**
   * Create a new liquidity pool
   * @throws ModuleDisabledError if DEX module is disabled
   */
  async createPool(
    creator: string,
    params: PoolParams,
    options?: GasOptions
  ): Promise<TxResult> {
    this.assertEnabled();

    const message = {
      typeUrl: '/paw.dex.v1.MsgCreatePool',
      value: {
        creator,
        tokenA: params.tokenA,
        tokenB: params.tokenB,
        amountA: params.amountA,
        amountB: params.amountB
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(creator, [message], options);
  }

  /**
   * Add liquidity to an existing pool
   * @throws ModuleDisabledError if DEX module is disabled
   */
  async addLiquidity(
    sender: string,
    params: AddLiquidityParams,
    options?: GasOptions
  ): Promise<TxResult> {
    this.assertEnabled();

    const message = {
      typeUrl: '/paw.dex.v1.MsgAddLiquidity',
      value: {
        sender,
        poolId: params.poolId,
        amountA: params.amountA,
        amountB: params.amountB,
        minShares: params.minShares
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(sender, [message], options);
  }

  /**
   * Remove liquidity from a pool
   * @throws ModuleDisabledError if DEX module is disabled
   */
  async removeLiquidity(
    sender: string,
    params: RemoveLiquidityParams,
    options?: GasOptions
  ): Promise<TxResult> {
    this.assertEnabled();

    const message = {
      typeUrl: '/paw.dex.v1.MsgRemoveLiquidity',
      value: {
        sender,
        poolId: params.poolId,
        shares: params.shares,
        minAmountA: params.minAmountA,
        minAmountB: params.minAmountB
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(sender, [message], options);
  }

  /**
   * Swap tokens
   * @throws ModuleDisabledError if DEX module is disabled
   */
  async swap(
    sender: string,
    params: SwapParams,
    options?: GasOptions
  ): Promise<TxResult> {
    this.assertEnabled();

    const message = {
      typeUrl: '/paw.dex.v1.MsgSwap',
      value: {
        sender,
        poolId: params.poolId,
        tokenIn: params.tokenIn,
        amountIn: params.amountIn,
        minAmountOut: params.minAmountOut,
        recipient: params.recipient || sender
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(sender, [message], options);
  }

  /**
   * Get pool by ID
   * Returns null when DEX module is disabled
   */
  async getPool(poolId: string): Promise<Pool | null> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/paw/dex/v1/pools/${poolId}`);
      if (!response.ok) {
        return null;
      }

      const data = await response.json() as { code?: number; pool?: Pool };
      // Check for "Not Implemented" response
      if (data.code === 12) {
        console.warn('DEX module is disabled in this network');
        return null;
      }
      return data.pool || null;
    } catch (error) {
      console.error('Error fetching pool:', error);
      return null;
    }
  }

  /**
   * Get all pools
   * Returns empty array when DEX module is disabled
   */
  async getAllPools(): Promise<Pool[]> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/paw/dex/v1/pools`);
      if (!response.ok) {
        return [];
      }

      const data = await response.json() as { code?: number; pools?: Pool[] };
      // Check for "Not Implemented" response
      if (data.code === 12) {
        console.warn('DEX module is disabled in this network');
        return [];
      }
      return data.pools || [];
    } catch (error) {
      console.error('Error fetching pools:', error);
      return [];
    }
  }

  /**
   * Get pool for token pair
   * Returns null when DEX module is disabled
   */
  async getPoolByTokens(tokenA: string, tokenB: string): Promise<Pool | null> {
    const pools = await this.getAllPools();
    return pools.find(pool =>
      (pool.tokenA === tokenA && pool.tokenB === tokenB) ||
      (pool.tokenA === tokenB && pool.tokenB === tokenA)
    ) || null;
  }

  /**
   * Calculate swap output amount (offline calculation)
   */
  calculateSwapOutput(
    amountIn: string,
    reserveIn: string,
    reserveOut: string,
    swapFee: string = '0.003'
  ): string {
    const amountInBig = BigInt(amountIn);
    const reserveInBig = BigInt(reserveIn);
    const reserveOutBig = BigInt(reserveOut);
    const feeBig = BigInt(Math.floor(parseFloat(swapFee) * 10000));

    // Apply fee: amountInWithFee = amountIn * (10000 - fee) / 10000
    const amountInWithFee = (amountInBig * (10000n - feeBig)) / 10000n;

    // Constant product formula: amountOut = (amountInWithFee * reserveOut) / (reserveIn + amountInWithFee)
    const numerator = amountInWithFee * reserveOutBig;
    const denominator = reserveInBig + amountInWithFee;

    return (numerator / denominator).toString();
  }

  /**
   * Calculate price impact (offline calculation)
   */
  calculatePriceImpact(
    amountIn: string,
    reserveIn: string,
    reserveOut: string
  ): number {
    const amountOut = this.calculateSwapOutput(amountIn, reserveIn, reserveOut, '0');
    const priceBeforeSwap = parseFloat(reserveOut) / parseFloat(reserveIn);
    const priceAfterSwap = parseFloat(amountOut) / parseFloat(amountIn);

    return Math.abs((priceAfterSwap - priceBeforeSwap) / priceBeforeSwap) * 100;
  }

  /**
   * Calculate shares for liquidity addition (offline calculation)
   */
  calculateShares(
    amountA: string,
    amountB: string,
    reserveA: string,
    _reserveB: string,
    totalShares: string
  ): string {
    if (totalShares === '0') {
      // First liquidity provider
      const amountABig = BigInt(amountA);
      const amountBBig = BigInt(amountB);
      return (amountABig * amountBBig).toString();
    }

    const amountABig = BigInt(amountA);
    const reserveABig = BigInt(reserveA);
    const totalSharesBig = BigInt(totalShares);

    return ((amountABig * totalSharesBig) / reserveABig).toString();
  }
}
